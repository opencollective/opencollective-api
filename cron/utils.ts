import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
