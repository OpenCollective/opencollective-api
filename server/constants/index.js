import activities from './activities';
import channels from './channels';
import expenseStatus from './expense_status';
import math from './math';
import roles from './roles';
import * as transactions from './transactions';

export { activities, expenseStatus, roles, math, channels, transactions };
