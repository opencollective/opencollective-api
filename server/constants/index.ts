import activities from './activities';
import channels from './channels';
import expenseStatus from './expense_status';
import expenseTypes from './expense_type';
import frequencies from './frequencies';
import math from './math';
import roles from './roles';
import * as transactions from './transactions';

export { activities, expenseStatus, expenseTypes, roles, math, channels, transactions, frequencies };
