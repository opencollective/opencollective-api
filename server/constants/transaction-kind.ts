export enum TransactionKind {
  /** Transactions coming from the "Add Funds" */
  ADDED_FUNDS = 'ADDED_FUNDS',
  /** Transactions from transferring the remaining balance from a project/event/collective **/
  BALANCE_TRANSFER = 'BALANCE_TRANSFER',
  /** Transactions coming from the "Contribution Flow" */
  CONTRIBUTION = 'CONTRIBUTION',
  /** Transactions coming from the "Expense Flow" */
  EXPENSE = 'EXPENSE',
  /** The host fee going to the Host */
  HOST_FEE = 'HOST_FEE',
  /** Part of the Host fee going from the Host to the Platform */
  HOST_FEE_SHARE = 'HOST_FEE_SHARE',
  /** Part of the Host fee going from the Host to the Platform */
  HOST_FEE_SHARE_DEBT = 'HOST_FEE_SHARE_DEBT',
  /** Amount given by Fiscal Hosts to cover payment processor fee on refunds */
  PAYMENT_PROCESSOR_COVER = 'PAYMENT_PROCESSOR_COVER',
  /** Reserved keyword in case we want to use in the future */
  PAYMENT_PROCESSOR_FEE = 'PAYMENT_PROCESSOR_FEE',
  /** Reserved keyword in case we want to use in the future */
  PLATFORM_FEE = 'PLATFORM_FEE',
  /** Financial contribution to Open Collective added on top of another contribution */
  PLATFORM_TIP = 'PLATFORM_TIP',
  /** Financial contribution to Open Collective added on top of another contribution */
  PLATFORM_TIP_DEBT = 'PLATFORM_TIP_DEBT',
  /** For prepaid budgets */
  PREPAID_PAYMENT_METHOD = 'PREPAID_PAYMENT_METHOD',
}
