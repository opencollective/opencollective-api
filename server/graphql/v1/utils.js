import debugLib from 'debug';
import { graphql } from 'graphql';

import { loaders } from '../loaders';

import schema from './schema';

const debug = debugLib('graphql');

export const makeRequest = (remoteUser, query) => {
  return {
    remoteUser,
    body: { query },
    loaders: loaders({ remoteUser }),
  };
};

export const graphqlQuery = async (query, variables, remoteUser) => {
  const prepare = () => {
    if (remoteUser) {
      remoteUser.rolesByCollectiveId = null; // force refetching the roles
      return remoteUser.populateRoles();
    } else {
      return Promise.resolve();
    }
  };

  if (process.env.DEBUG && process.env.DEBUG.match(/graphql/)) {
    debug('query', query);
    debug('variables', variables);
    debug('context', remoteUser);
  }

  return prepare().then(() =>
    graphql({
      schema,
      source: query,
      rootValue: null,
      contextValue: makeRequest(remoteUser, query),
      variableValues: variables,
    }),
  );
};
