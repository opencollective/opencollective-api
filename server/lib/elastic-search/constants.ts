export enum ElasticSearchIndexName {
  COLLECTIVES = 'collectives',
  COMMENTS = 'comments',
  EXPENSES = 'expenses',
  UPDATES = 'updates',
  TRANSACTIONS = 'transactions',
  ORDERS = 'orders',
  TIERS = 'tiers',
  HOST_APPLICATIONS = 'host-applications',
}
