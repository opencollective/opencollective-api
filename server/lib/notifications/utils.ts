import sanitizeHtml from 'sanitize-html';

import { sanitizerOptions as updateSanitizerOptions } from '../../models/Update';

const constructPreviewImageURL = (service: string, id: string) => {
  if (service === 'youtube' && id.match('[a-zA-Z0-9_-]{11}')) {
    return `https://img.youtube.com/vi/${id}/0.jpg`;
  } else if (service === 'anchorFm') {
    return `https://opencollective.com/static/images/anchor-fm-logo.png`;
  } else {
    return null;
  }
};

const parseServiceLink = (videoLink: string) => {
  const regexps = {
    youtube: new RegExp(
      '(?:https?://)?(?:www\\.)?youtu(?:\\.be/|be(-nocookie)?\\.com/\\S*(?:watch|embed)(?:(?:(?=/[^&\\s?]+(?!\\S))/)|(?:\\S*v=|v/)))([^&\\s?]+)',
      'i',
    ),
    anchorFm: /^(http|https)?:\/\/(www\.)?anchor\.fm\/([^/]+)(\/embed)?(\/episodes\/)?([^/]+)?\/?$/,
  };
  for (const service in regexps) {
    videoLink = videoLink.replace('/?showinfo=0', '');
    const matches = regexps[service].exec(videoLink);
    if (matches) {
      if (service === 'anchorFm') {
        const podcastName = matches[3];
        const episodeId = matches[6];
        const podcastUrl = `${podcastName}/embed`;
        return { service, id: episodeId ? `${podcastUrl}/episodes/${episodeId}` : podcastUrl };
      } else {
        return { service, id: matches[matches.length - 1] };
      }
    }
  }
  return {};
};

export const replaceVideosByImagePreviews = (html: string) => {
  const sanitizerOptions = {
    ...updateSanitizerOptions,
    transformTags: {
      ...updateSanitizerOptions.transformTags,
      iframe: (tagName, attribs) => {
        if (!attribs.src) {
          return '';
        }
        const { service, id } = parseServiceLink(attribs.src);
        const imgSrc = constructPreviewImageURL(service, id);
        if (imgSrc) {
          return {
            tagName: 'img',
            attribs: {
              src: imgSrc,
              alt: `${service} content`,
            },
          };
        } else {
          return '';
        }
      },
    },
  };
  return sanitizeHtml(html, sanitizerOptions);
};
