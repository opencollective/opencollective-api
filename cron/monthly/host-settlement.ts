#!/usr/bin/env node
import '../../server/env';

import config from 'config';
import { parse as json2csv } from 'json2csv';
import { groupBy, pick, sumBy } from 'lodash';
import moment from 'moment';
import { v4 as uuid } from 'uuid';

import activityType from '../../server/constants/activities';
import expenseStatus from '../../server/constants/expense_status';
import expenseTypes from '../../server/constants/expense_type';
import { SETTLEMENT_EXPENSE_PROPERTIES } from '../../server/constants/transactions';
import { uploadToS3 } from '../../server/lib/awsS3';
import { getPendingHostFeeShare, getPendingPlatformTips } from '../../server/lib/host-metrics';
import { parseToBoolean } from '../../server/lib/utils';
import models, { sequelize } from '../../server/models';
import { PayoutMethodTypes } from '../../server/models/PayoutMethod';

const today = moment.utc();

const defaultDate = process.env.START_DATE ? moment.utc(process.env.START_DATE) : moment.utc();

const DRY = process.env.DRY;
const HOST_ID = process.env.HOST_ID;
const isProduction = config.env === 'production';

// Only run on the 1th of the month
if (isProduction && new Date().getDate() !== 1 && !process.env.OFFCYCLE) {
  console.log('OC_ENV is production and today is not the 1st of month, script aborted!');
  process.exit();
} else if (parseToBoolean(process.env.SKIP_HOST_SETTLEMENT)) {
  console.log('Skipping because SKIP_HOST_SETTLEMENT is set.');
  process.exit();
}

if (DRY) {
  console.info('Running dry, changes are not going to be persisted to the DB.');
}

const ATTACHED_CSV_COLUMNS = ['createdAt', 'description', 'amount', 'currency', 'OrderId', 'TransactionGroup'];

export async function run(baseDate: Date | moment.Moment = defaultDate): Promise<void> {
  const momentDate = moment(baseDate).subtract(1, 'month');
  const year = momentDate.year();
  const month = momentDate.month();
  const startDate = new Date(year, month, 1);
  const endDate = new Date(year, month + 1, 1);

  console.info(`Invoicing hosts pending fees and tips for ${momentDate.format('MMMM')}.`);

  const payoutMethods = groupBy(
    await models.PayoutMethod.findAll({
      where: { CollectiveId: SETTLEMENT_EXPENSE_PROPERTIES.FromCollectiveId, isSaved: true },
    }),
    'type',
  );

  const hosts = await sequelize.query(
    `
      SELECT c.*
      FROM "Collectives" c
      INNER JOIN "Transactions" t ON t."HostCollectiveId" = c.id AND t."deletedAt" IS NULL
      WHERE c."isHostAccount" IS TRUE
      AND t."createdAt" >= :startDate AND t."createdAt" < :endDate
      AND c.id != 8686 -- Make sure we don't invoice OC Inc as reverse settlements are not supported yet
      GROUP BY c.id
    `,
    {
      mapToModel: true,
      type: sequelize.QueryTypes.SELECT,
      model: models.Collective,
      replacements: { startDate: startDate, endDate: endDate },
    },
  );

  for (const host of hosts) {
    const pendingPlatformTips = await getPendingPlatformTips(host, { startDate, endDate });
    const pendingHostFeeShare = await getPendingHostFeeShare(host, { startDate, endDate });

    if (HOST_ID && host.id !== parseInt(HOST_ID)) {
      continue;
    }

    const plan = await host.getPlan();

    let items = [];

    const transactions = await sequelize.query(
      `SELECT t.*
FROM "Transactions" as t
INNER JOIN "TransactionSettlements" ts ON ts."TransactionGroup" = t."TransactionGroup" AND t.kind = ts.kind
WHERE t."CollectiveId" = :CollectiveId
AND t."createdAt" >= :startDate AND t."createdAt" < :endDate
AND t."kind" IN ('PLATFORM_TIP_DEBT', 'HOST_FEE_SHARE_DEBT')
AND t."isDebt" IS TRUE
AND t."deletedAt" IS NULL
AND ts."status" != 'SETTLED'`,
      {
        replacements: { CollectiveId: host.id, startDate: startDate, endDate: endDate },
        model: models.Transaction,
        mapToModel: true, // pass true here if you have any mapped fields
      },
    );

    items.push({
      incurredAt: new Date(),
      amount: pendingPlatformTips,
      description: 'Platform Tips',
    });

    items.push({
      incurredAt: new Date(),
      amount: pendingHostFeeShare,
      description: 'Shared Revenue',
    });

    if (plan.pricePerCollective) {
      const activeHostedCollectives = await host.getHostedCollectivesCount();
      const amount = (activeHostedCollectives || 0) * plan.pricePerCollective;
      if (amount) {
        items.push({
          incurredAt: new Date(),
          amount,
          description: 'Fixed Fee per Hosted Collective',
        });
      }
    }

    const totalAmountCharged = sumBy(items, 'amount');

    if (totalAmountCharged < 1000) {
      console.warn(
        `${host.name} (#${host.id}) skipped, total amount pending ${totalAmountCharged / 100} < 10.00 ${
          host.currency
        }.\n`,
      );
      continue;
    }
    console.info(
      `${host.name} (#${host.id}) has ${transactions.length} pending transactions and owes ${
        totalAmountCharged / 100
      } (${host.currency})`,
    );

    let csv;
    if (transactions.length) {
      csv = json2csv(transactions.map(t => pick(t, ATTACHED_CSV_COLUMNS)));
    }

    if (DRY) {
      console.debug(`Items:\n${json2csv(items)}\n`);
    } else {
      const connectedAccounts = await host.getConnectedAccounts({
        where: { deletedAt: null },
      });

      let payoutMethod =
        payoutMethods[PayoutMethodTypes.OTHER]?.[0] || payoutMethods[PayoutMethodTypes.BANK_ACCOUNT]?.[0];
      if (
        connectedAccounts?.find(c => c.service === 'transferwise') &&
        payoutMethods[PayoutMethodTypes.BANK_ACCOUNT]?.[0]
      ) {
        const currencyCompatibleAccount = payoutMethods[PayoutMethodTypes.BANK_ACCOUNT].find(
          pm => pm.data?.['currency'] === host.currency,
        );
        payoutMethod = currencyCompatibleAccount || payoutMethods[PayoutMethodTypes.BANK_ACCOUNT]?.[0];
      } else if (
        connectedAccounts?.find(c => c.service === 'paypal') &&
        !host.settings?.disablePaypalPayouts &&
        payoutMethods[PayoutMethodTypes.PAYPAL]?.[0]
      ) {
        payoutMethod = payoutMethods[PayoutMethodTypes.PAYPAL]?.[0];
      }

      if (!payoutMethod) {
        console.error('No Payout Method found, Open Collective Inc. needs to have at least one payout method.');
        process.exit();
      }

      // Create the Expense
      const transactionIds = transactions.map(t => t.TransactionId);
      const expense = await models.Expense.create({
        ...SETTLEMENT_EXPENSE_PROPERTIES,
        PayoutMethodId: payoutMethod.id,
        amount: totalAmountCharged,
        CollectiveId: host.id,
        currency: host.currency,
        description: `Platform settlement for ${momentDate.utc().format('MMMM')}`,
        incurredAt: today,
        data: { isPlatformTipSettlement: true, transactionIds },
        type: expenseTypes.INVOICE,
        status: expenseStatus.PENDING,
      });

      // Create Expense Items
      items = items.map(i => ({
        ...i,
        ExpenseId: expense.id,
        CreatedByUserId: SETTLEMENT_EXPENSE_PROPERTIES.UserId,
      }));

      await models.ExpenseItem.bulkCreate(items);

      // Attach CSV
      if (csv) {
        const Body = csv;
        const filenameBase = `${host.name}-${momentDate.format('MMMM-YYYY')}`;
        const Key = `${filenameBase}.${uuid().split('-')[0]}.csv`;
        const { Location: url } = await uploadToS3({
          Bucket: config.aws.s3.bucket,
          Key,
          Body,
          ACL: 'public-read',
          ContentType: 'text/csv',
        });
        await models.ExpenseAttachedFile.create({
          url,
          ExpenseId: expense.id,
          CreatedByUserId: SETTLEMENT_EXPENSE_PROPERTIES.UserId,
        });
      }

      // Mark transactions as invoiced
      await models.TransactionSettlement.markTransactionsAsInvoiced(transactions, expense.id);
      await expense.createActivity(activityType.COLLECTIVE_EXPENSE_CREATED);
    }
  }
}

if (require.main === module) {
  run(defaultDate)
    .catch(e => {
      console.error(e);
      process.exit(1);
    })
    .then(() => {
      process.exit();
    });
}
