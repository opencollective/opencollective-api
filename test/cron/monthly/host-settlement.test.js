import { expect } from 'chai';
import moment from 'moment';

import { run as invoicePlatformFees } from '../../../cron/monthly/host-settlement';
import { TransactionKind } from '../../../server/constants/transaction-kind';
import models, { sequelize } from '../../../server/models';
import { TransactionSettlementStatus } from '../../../server/models/TransactionSettlement';
import {
  fakeCollective,
  fakeConnectedAccount,
  fakeHost,
  fakePaymentMethod,
  fakePayoutMethod,
  fakeTransaction,
  fakeUser,
} from '../../test-helpers/fake-data';
import * as utils from '../../utils';

describe('cron/monthly/host-settlement', () => {
  const lastMonth = moment.utc().subtract(1, 'month');

  let gbpHost, expense;
  before(async () => {
    await utils.resetTestDB();
    const user = await fakeUser({ id: 30 }, { id: 20, slug: 'pia' });
    const oc = await fakeHost({ id: 8686, slug: 'opencollective', CreatedByUserId: user.id });

    // Move Collectives ID auto increment pointer up, so we don't collide with the manually created id:1
    await sequelize.query(`ALTER SEQUENCE "Collectives_id_seq" RESTART WITH 1453`);
    const payoutProto = {
      data: {
        details: {},
        type: 'IBAN',
        accountHolderName: 'OpenCollective Inc.',
        currency: 'USD',
      },
      CollectiveId: oc.id,
      type: 'BANK_ACCOUNT',
    };
    await fakePayoutMethod({
      ...payoutProto,
      id: 2955,
    });
    await fakePayoutMethod({
      ...payoutProto,
      id: 2956,
      data: { ...payoutProto.data, currency: 'GBP' },
    });

    gbpHost = await fakeHost({ currency: 'GBP', plan: 'grow-plan-2021', data: { plan: { pricePerCollective: 100 } } });
    await fakeConnectedAccount({ CollectiveId: gbpHost.id, service: 'transferwise' });

    const socialCollective = await fakeCollective({ HostCollectiveId: gbpHost.id });
    const transactionProps = {
      type: 'CREDIT',
      kind: TransactionKind.CONTRIBUTION,
      CollectiveId: socialCollective.id,
      currency: 'GBP',
      hostCurrency: 'GBP',
      HostCollectiveId: gbpHost.id,
      createdAt: lastMonth,
    };
    // Create Contributions
    const contribution1 = await fakeTransaction({
      ...transactionProps,
      kind: TransactionKind.CONTRIBUTION,
      amount: 3000,
      hostFeeInHostCurrency: -600,
    });
    const contribution2 = await fakeTransaction({
      ...transactionProps,
      kind: TransactionKind.CONTRIBUTION,
      amount: 3000,
      hostFeeInHostCurrency: -400,
    });
    const contribution3 = await fakeTransaction({
      ...transactionProps,
      kind: TransactionKind.CONTRIBUTION,
      amount: 3000,
      hostFeeInHostCurrency: -600,
    });
    // Create host fee share
    const hostFeeResults = await Promise.all(
      [contribution1, contribution2, contribution3].map(transaction =>
        models.Transaction.createHostFeeTransactions(transaction, gbpHost),
      ),
    );

    await Promise.all(
      hostFeeResults.map(({ transaction, hostFeeTransaction }) =>
        models.Transaction.createHostFeeShareTransactions(
          { transaction: transaction, hostFeeTransaction: hostFeeTransaction },
          gbpHost,
          false,
        ),
      ),
    );

    // Add Platform Tips
    const t = await fakeTransaction(transactionProps);
    await fakeTransaction({
      type: 'CREDIT',
      CollectiveId: oc.id,
      HostCollectiveId: oc.id,
      amount: 1000,
      currency: 'USD',
      data: { hostToPlatformFxRate: 1.23 },
      TransactionGroup: t.TransactionGroup,
      kind: TransactionKind.PLATFORM_TIP,
      createdAt: lastMonth,
    });
    const firstTipDebtCredit = await fakeTransaction({
      type: 'CREDIT',
      FromCollectiveId: oc.id,
      CollectiveId: gbpHost.id,
      HostCollectiveId: gbpHost.id,
      amount: 813,
      amountInHostCurrency: 813,
      currency: 'GBP',
      hostCurrency: 'GBP',
      data: { hostToPlatformFxRate: 1.23 },
      TransactionGroup: t.TransactionGroup,
      kind: TransactionKind.PLATFORM_TIP_DEBT,
      createdAt: lastMonth,
      isDebt: true,
    });
    await models.TransactionSettlement.createForTransaction(firstTipDebtCredit);

    // Collected Platform Tip with pending Payment Processor Fee. No debt here, it's collected directly via Stripe
    const t2 = await fakeTransaction(transactionProps);
    const paymentMethod = await fakePaymentMethod({ service: 'stripe', token: 'tok_bypassPending' });
    await fakeTransaction({
      type: 'CREDIT',
      CollectiveId: oc.id,
      HostCollectiveId: oc.id,
      amount: 813,
      amountInHostCurrency: 813,
      hostCurrency: 'GBP',
      data: { hostToPlatformFxRate: 1.23 },
      TransactionGroup: t2.TransactionGroup,
      kind: TransactionKind.PLATFORM_TIP,
      paymentProcessorFeeInHostCurrency: -100,
      PaymentMethodId: paymentMethod.id,
      createdAt: lastMonth,
    });

    await invoicePlatformFees();

    expense = (await gbpHost.getExpenses())[0];
    expect(expense).to.exist;
    expense.items = await expense.getItems();
  });

  // Resync DB to make sure we're not touching other tests
  after(async () => {
    await utils.resetTestDB();
  });

  it('should invoice the host in its own currency', () => {
    expect(expense).to.have.property('currency', 'GBP');
    expect(expense).to.have.property('description').that.includes('Platform settlement for');
    expect(expense).to.have.nested.property('data.isPlatformTipSettlement', true);
  });

  it('should use a payout method compatible with the host currency', () => {
    expect(expense).to.have.property('PayoutMethodId', 2956);
  });

  it('should invoice platform tips not collected through Stripe', async () => {
    const platformTipsItem = expense.items.find(p => p.description === 'Platform Tips');
    expect(platformTipsItem).to.have.property('amount', Math.round(1000 / 1.23));
  });

  it('should invoice pending shared host revenue', async () => {
    const sharedRevenueItem = expense.items.find(p => p.description === 'Shared Revenue');
    expect(sharedRevenueItem).to.have.property('amount', Math.round(1600 * 0.15));
  });

  it('should attach detailed list of transactions in the expense', async () => {
    const [attachment] = await expense.getAttachedFiles();
    expect(attachment).to.have.property('url').that.includes('.csv');
  });

  it('should consider fixed fee per host collective', async () => {
    const reimburseItem = expense.items.find(p => p.description === 'Fixed Fee per Hosted Collective');
    expect(reimburseItem).to.have.property('amount', 100);
  });

  it('should update settlementStatus to INVOICED', async () => {
    const settlements = await models.TransactionSettlement.findAll();
    expect(settlements.length).to.eq(4); // 1 Platform tip + 3 host fee share
    settlements.forEach(settlement => {
      expect(settlement.status).to.eq(TransactionSettlementStatus.INVOICED);
    });
  });
});
