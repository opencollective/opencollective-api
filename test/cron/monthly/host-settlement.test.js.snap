// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cron/monthly/host-settlement should update all settlement status for EU host 1`] = `
"
| kind                | type   | amount | isRefund | Settlement |
| ------------------- | ------ | ------ | -------- | ---------- |
| PLATFORM_TIP_DEBT   | CREDIT | 30000  | false    | INVOICED   |
| HOST_FEE_SHARE_DEBT | CREDIT | 5000   | false    | INVOICED   |"
`;

exports[`cron/monthly/host-settlement should update all settlement status for GBP host 1`] = `
"
| Group     | kind                | type   | amount | isRefund | Settlement |
| --------- | ------------------- | ------ | ------ | -------- | ---------- |
| #00000001 | HOST_FEE_SHARE_DEBT | CREDIT | 90     | false    | INVOICED   |
| #00000002 | HOST_FEE_SHARE_DEBT | CREDIT | 60     | false    | INVOICED   |
| #00000003 | HOST_FEE_SHARE_DEBT | CREDIT | 90     | false    | INVOICED   |
| #00000004 | HOST_FEE_SHARE_DEBT | CREDIT | 90     | false    | SETTLED    |
| #00000005 | HOST_FEE_SHARE_DEBT | CREDIT | 63     | false    | SETTLED    |
| #00000006 | PLATFORM_TIP_DEBT   | CREDIT | 813    | false    | INVOICED   |
| #00000008 | HOST_FEE_SHARE_DEBT | DEBIT  | -90    | true     | SETTLED    |
| #00000009 | HOST_FEE_SHARE_DEBT | DEBIT  | -63    | true     | INVOICED   |"
`;
